### DISCUSSION TITLE
has anyone of you guys injected Claude Code system tools into llama.cpp cache

### DISCUSSION THREAD

**Discussion by aliaj00** [ggml-org/llama.cpp#19467] [General]
hi guys,

I have a hp laptop with AMD and 32GB Ram. i compiled the llama.cpp on Debian 12 and i am running qwen3-coder on it with approx 4 tokens/sec.  I am trying to get Claude code to work with llama.cpp and qwen3-coder, but the issue is that Claude in the beginning send at least 2 big JSON with approx 16-17k tokens, and it clogs the llama.cpp and make Claude code timeout. After I inject the JSON manually via curl, it start working at the speed that llama server can process 27 token/sec analyzing and 4 token//sec decoding.

Has anyone managed to get the system and tools prompts from Claude and inject into cache successfully so it hit them and does not rework them. I am able to inject and work but as it adds the time stamp and user_id kind of hash makes that cache invalid for the next day or session. So far iI was able to save and load cache for the same session.

==============llama.cpp server partial logs=============
srv  log_server_r: request: POST /v1/messages 127.0.0.1 200
srv   prompt_save:  - saving prompt with length 122, total state size = 11.440 MiB
srv          load:  - looking for better prompt, base f_keep = 0.123, sim = 0.001
srv        update:  - cache state: 4 prompts, 4436.796 MiB (limits: 0.000 MiB, 64000 tokens, 64000 est)
srv        update:    - prompt 0x5641daf38d40:   16786 tokens, checkpoints:  0,  1573.881 MiB
srv        update:    - prompt 0x5641dadf9b50:   13578 tokens, checkpoints:  0,  1273.094 MiB
srv        update:    - prompt 0x5641da650690:   16834 tokens, checkpoints:  0,  1578.381 MiB
srv        update:    - prompt 0x5641daafd580:     122 tokens, checkpoints:  0,    11.440 MiB
srv  get_availabl: prompt cache update took 5.44 ms
slot launch_slot_: id  0 | task -1 | sampler chain: logits -> logit-bias -> penalties -> dry -> top-n-sigma -> top-k -> typical -> top-p -> min-p -> xtc -> temp-ext -> dist 
slot launch_slot_: id  0 | task 122 | processing task
slot update_slots: id  0 | task 122 | new prompt, n_ctx_slot = 64000, n_keep = 0, task.n_tokens = 16933
slot update_slots: id  0 | task 122 | n_tokens = 15, memory_seq_rm [15, end)
slot update_slots: id  0 | task 122 | prompt processing progress, n_tokens = 1039, batch.n_tokens = 1024, progress = 0.061359
slot update_slots: id  0 | task 122 | n_tokens = 1039, memory_seq_rm [1039, end)
slot update_slots: id  0 | task 122 | prompt processing progress, n_tokens = 2063, batch.n_tokens = 1024, progress = 0.121833
slot update_slots: id  0 | task 122 | n_tokens = 2063, memory_seq_rm [2063, end)
slot update_slots: id  0 | task 122 | prompt processing progress, n_tokens = 3087, batch.n_tokens = 1024, progress = 0.182307
slot update_slots: id  0 | task 122 | n_tokens = 3087, memory_seq_rm [3087, end)
slot update_slots: id  0 | task 122 | prompt processing progress, n_tokens = 4111, batch.n_tokens = 1024, progress = 0.242780
slot update_slots: id  0 | task 122 | n_tokens = 4111, memory_seq_rm [4111, end)
slot update_slots: id  0 | task 122 | prompt processing progress, n_tokens = 5135, batch.n_tokens = 1024, progress = 0.303254
srv          stop: cancel task, id_task = 122

====================================================
the working after processing the firt reqests
 
srv  params_from_: Chat format: Qwen3 Coder
slot get_availabl: id  0 | task -1 | selected slot by LCP similarity, sim_best = 0.998 (> 0.100 thold), f_keep = 0.999
slot launch_slot_: id  0 | task -1 | sampler chain: logits -> logit-bias -> penalties -> dry -> top-n-sigma -> top-k -> typical -> top-p -> min-p -> xtc -> temp-ext -> dist 
slot launch_slot_: id  0 | task 291 | processing task
slot update_slots: id  0 | task 291 | new prompt, n_ctx_slot = 64000, n_keep = 0, task.n_tokens = 17046
slot update_slots: id  0 | task 291 | n_tokens = 17015, memory_seq_rm [17015, end)
slot update_slots: id  0 | task 291 | prompt processing progress, n_tokens = 17046, batch.n_tokens = 31, progress = 1.000000
slot update_slots: id  0 | task 291 | prompt done, n_tokens = 17046, batch.n_tokens = 31
slot print_timing: id  0 | task 291 | 
prompt eval time =   20370.40 ms /    31 tokens (  657.11 ms per token,     1.52 tokens per second)
       eval time =   95302.00 ms /    93 tokens ( 1024.75 ms per token,     0.98 tokens per second)
      total time =  115672.40 ms /   124 tokens

I have been trying to prime it to the llama.cpp before hand but the info on the JSON has the timestamp on the system and tool prompts and extra data  like:

================
"Today's date is 2026-02-07"
================
"metadata": {
    "user_id": "user_f68009ad6fb6e4aa85edb591c9bb23a0c53fee05575fdf9e58a8ac96ae042808_account__session_45ec8643-2e8d-496d-9aee-5bd3fee257e7"
  },
================
